### PR TITLE
Fix nodejs_image layering

### PR DESCRIPTION
**Pull request by Globegitter**
This builds upon #342 and fixes the layering for the nodejs_image making it work with the latest rules. This is dependent on https://github.com/bazelbuild/rules_nodejs/pull/283 being merged into rules_nodejs.

The top layer of the example nodejs_image goes down from 64MB to ~60KB. From one of our own apps it reduces it from ~223MB to ~28MB

### CODE REVIEW

File: lang/image.bzl [hunk 1]
```diff
@@ -224,6 +226,7 @@ def _app_layer_impl(ctx, runfiles = None, emptyfiles = None):
         directory = "/",
         file_map = file_map,
         empty_files = empty_files,
+        empty_dirs = emptydirs,
```
Comment (nlopezgi) [inline lang/image.bzl]: Shouldn't we perform checks on these emptydirs (as is done for emptyfiles in lines 195-199) to verify there are no collisions between passed args and other files/dirs?

File: nodejs/image.bzl [hunk 2]
```diff
@@ -59,6 +61,14 @@ DEFAULT_BASE = select({
     "//conditions:default": "@nodejs_debug_image_base//image",
 })
 
+load("@build_bazel_rules_nodejs//:defs.bzl", "SourcesInfo")
```
Comment (nlopezgi) [inline nodejs/image.bzl]: can you alias this with a name that is more descriptive of this being specific a nodejs definition (could you also please comment on what it does in _app_runfiles below?)

File: nodejs/image.bzl [hunk 3]
```diff
@@ -18,7 +18,9 @@ The signature of this rule is compatible with nodejs_binary.
 
 load(
     "//lang:image.bzl",
-    "app_layer",
+    "app_layer_attrs",
+    "app_layer_impl",
```
Comment (nlopezgi) [inline nodejs/image.bzl]: can you alias these with lang_ so that they are visibly distinct from the ones declared here

File: nodejs/image.bzl [hunk 4]
```diff
@@ -69,34 +79,23 @@ def _dep_layer_impl(ctx):
...
-        "directory": attr.string(default = "/app"),
-    }.items()),
+    attrs = dep_layer_attrs,
     executable = True,
     outputs = _container.image.outputs,
     implementation = _dep_layer_impl,
 )
 
+def _app_layer_impl(ctx):
+    empty_dirs = ["/".join(["/app", ctx.workspace_name])]
```
Comment (nlopezgi) [inline nodejs/image.bzl]: can you document here briefly why the empty dirs need to be added (and anything else specific about this impl that makes it different from the one in lang/image.bzl, e.g., what you are doing with the runfiles)

File: nodejs/image.bzl [hunk 5]
```diff
@@ -114,13 +113,7 @@ def nodejs_image(
...
   """
     binary_name = name + ".binary"
-
-    layers = [
-        # Put the Node binary into its own layer.
-        "@nodejs//:node",
-        # node_modules can get large, it should be in its own layer.
-        node_modules,
-    ] + layers
+    node_binary_name = binary_name + "_bin"
```
Comment (nlopezgi) [inline nodejs/image.bzl]: could you document why we need this additional var

File: nodejs/image.bzl [hunk 6]
```diff
@@ -129,21 +122,30 @@ def nodejs_image(
         **kwargs
     )
 
+    layers = [
```
Comment (nlopezgi) [inline nodejs/image.bzl]: any reason this now is after the nodejs_binary call above (if so, could you briefly document)

File: container/image_test.py [hunk 7]
```diff
@@ -575,6 +581,110 @@ def test_war_image(self):
       ])
 
 
+  def test_nodejs_image(self):
+    self.maxDiff = None
+    with TestImage('nodejs_image') as img:
+      # Check the application layer, which is on top, it also contains symlinks to the bottom layers.
```
Comment (nlopezgi) [inline container/image_test.py]: r/ it

**Comment by Globegitter** [comment]
Just as a note, I am waiting to get feedback on the `rules_nodejs` PR first before updating here, to make sure I don't have to update this PR multiple times.

**Comment by nlopezgi** [comment]
Hi @Globegitter any updates on this PR. We're close to accepting a PR that touches a lot of the same code as this PR (https://github.com/bazelbuild/rules_docker/pull/324) and also affects nodejs layering (please comment on that PR if you have any thoughts). Thanks

**Comment by Globegitter** [comment]
@nlopezgi Thanks for pinging me, unfortuantely review on the rules_nodejs has been slow, let me ping the maintainer again and I will check that other PR.

**Comment by Globegitter** [comment]
@nlopezgi It should be fine merging the other PR and I will just deal with any conflicts once I get back to this.

**Comment by nlopezgi** [comment]
hi @Globegitter 
is this PR still something you want plan on working on or can we close it otherwise?

**Comment by Toxicable** [comment]
We're still very much interested in this change.
It looks like the rules_nodejs PR is pretty much ready to go, just needs some additional tests and docs.

**Comment by Globegitter** [comment]
@nlopezgi I am sorry I dropped the ball on this I just have not had the time to give this any priority in the past months. Thanks for reminding me about this, I will put it on my todo again.

**Comment by Globegitter** [comment]
googlecloud is failing - is that remote execution? Has that ever worked with the nodejs images? If it is something I broke is it possible to see the logs?

**Comment by Globegitter** [comment]
So cleaned everything up, seems we do not need the `emptydirs` logic anymore so that is nice.

**Comment by nlopezgi** [comment]
hi @Globegitter ,
any progress on this or should I close?

**Comment by Globegitter** [comment]
@nlopezgi It has been a long process but it is still alive. There was a blocking bug uncovered last minute in https://github.com/bazelbuild/rules_nodejs/pull/283 which has been resolved recently so now I just need to update that PR and then I can fix this up here. So sorry for the long delay on this but it has not been entirely my fault.

**Comment by nlopezgi** [comment]
@Globegitter thanks for the update
no problem, no hurry on our end, just wanted to make sure this is still relevant to keep open.

**Comment by nlopezgi** [comment]
any updates on this PR, still relevant or should we close?

**Comment by Globegitter** [comment]
Yeah let's close it for now - I now have a new proposal to solve this: https://github.com/bazelbuild/rules_nodejs/pull/863 but still waiting for it to get merged. So once that i done I will create a new PR.

**Pull request closed without merging**